akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala

```diff
@@ -111,13 +118,16 @@ private[cluster] class ClusterRemoteWatcher(
     if (m.address != selfAddress) {
       quarantineOldIncarnation(m)
       clusterNodes += m.address
+      if (tombstones.contains(m.address)) // new node with same address joined
```

Comment:
As far as I can see we always publish MemberRemoved(old incarnation) before MemberJoined(new incarnation). For example when receiving a Gossip where the old is removed and new is joining.
In `diffMemberEvents`
```
val removedMembers = oldGossip.members.diff(newGossip.members)
```
and those are published first.

Maybe you can add one more test specifically for that case in ClusterDomainEventSpec.
We have something similar but not exactly that
"be produced for removed members"
"be produced for removed and rejoined member in another data center"

Reply:
I realise now the member diff creates both events, I moved the publish method because I was thinking it was two separate diffs and I needed to test it all.